### PR TITLE
Fix the issues with the common subexpression feature.

### PR DESCRIPTION
**Pull request by jianqiao**
This PR fixes the issues in #237.

### CODE REVIEW

File: query_optimizer/rules/ReuseAggregateExpressions.cpp [hunk 1]
```diff
@@ -194,7 +194,7 @@ P::PhysicalPtr ReuseAggregateExpressions::applyToNode(
             sum_it == ref_map.end() ? kInvalidRef : sum_it->second.front();
 
         for (const std::size_t idx : avg_it->second) {
-          agg_refs[idx].reset(new AggregateReference(sum_ref, count_ref));
+          agg_refs[idx] = std::make_unique<AggregateReference>(sum_ref, count_ref);
```
Comment (cramja) [inline query_optimizer/rules/ReuseAggregateExpressions.cpp]: What is the difference between using the assignment and `make_unique` versus `reset` and `new`?

Comment (zuyu) [inline query_optimizer/rules/ReuseAggregateExpressions.cpp]: Memory leaks upon system failures when using `new`.

**Comment by zuyu** [comment]
@jianqiao We missed the following comments in this PR:

* [Remove unused `moveTypedValue` method in `types/containers/ColumnVector.hpp`](https://github.com/apache/incubator-quickstep/pull/237#discussion_r112827764)
* [Change `hash_cache_` as a `std::uint64_t`, instead of a smart pointer](https://github.com/apache/incubator-quickstep/pull/237#discussion_r112828399)
* [`override` destructor needed for `CommonSubexpression`](https://github.com/apache/incubator-quickstep/pull/237#discussion_r112847577)

File: query_optimizer/rules/ExtractCommonSubexpression.hpp [hunk 2]
```diff
@@ -111,7 +111,7 @@ class ExtractCommonSubexpression : public Rule<physical::Physical> {
   bool visitAndCount(
       const expressions::ExpressionPtr &expression,
       ScalarCounter *counter,
-      ScalarHashable *hashable);
+      ScalarHashable *hashable) const;
```
Comment (zuyu) [inline query_optimizer/rules/ExtractCommonSubexpression.hpp]: All methods below in this file could be marked as `const`:

* `transformExpressions`
* `transformExpression`
* `visitAndTransform`

Comment (jianqiao) [inline query_optimizer/rules/ExtractCommonSubexpression.hpp]: These three methods involve the invoking of `optimizer_context_->nextExprId()` which modifies  the context (symbol id counter) of `optimizer_context_`.

Comment (zuyu) [inline query_optimizer/rules/ExtractCommonSubexpression.hpp]: Good catch!

**Pull request merged**